compiler/src/iree/compiler/Dialect/LinalgExt/Utils/MatchUtils.cpp

```diff
@@ -0,0 +1,329 @@
+      SmallVector<unsigned, 2>(batches.begin(), batches.end()),
+      SmallVector<unsigned, 2>(ac.begin(), ac.end()),
+      SmallVector<unsigned, 2>(bc.begin(), bc.end()),
+      SmallVector<unsigned, 2>(sa.begin(), sa.end()),
+      SmallVector<unsigned, 2>(ra.begin(), ra.end())};
+  llvm::sort(dimensions.batch.begin(), dimensions.batch.end());
+  llvm::sort(dimensions.m.begin(), dimensions.m.end());
+  llvm::sort(dimensions.n.begin(), dimensions.n.end());
+  llvm::sort(dimensions.k.begin(), dimensions.k.end());
+  llvm::sort(dimensions.kB.begin(), dimensions.kB.end());
+  return dimensions;
+}
+
+FailureOr<ScaledContractionDimensions>
+inferScaledContractionDims(ArrayRef<AffineMap> indexingMaps) {
+  if (indexingMaps.size() != 5)
+    return failure();
+  auto iterators = inferIteratorsFromOutMap(indexingMaps[4]);
+  if (failed(iterators))
+    return failure();
+  return inferScaledContractionDimsImpl(indexingMaps, iterators.value());
+}
+
+FailureOr<ScaledContractionDimensions>
+inferScaledContractionDims(linalg::LinalgOp linalgOp) {
+  if (linalgOp.getNumDpsInits() != 1 || linalgOp.getNumDpsInputs() != 4)
+    return failure();
+  return inferScaledContractionDimsImpl(linalgOp.getIndexingMapsArray(),
+                                        linalgOp.getIteratorTypesArray());
+}
+
+detail::MatchContractionResult
+isScaledContractionImpl(Operation *op,
+                        ScaledContractionDimensions *dimensions) {
+  auto linalgOp = dyn_cast<linalg::LinalgOp>(op);
+  if (!linalgOp)
+    return detail::MatchContractionResult::NotLinalgOp;
+  if (linalgOp.getNumDpsInputs() != 4 || linalgOp.getNumDpsInits() != 1)
+    return detail::MatchContractionResult::WrongNumOperands;
+  auto mapRange = linalgOp.getIndexingMapsArray();
+  if (linalgOp.getNumReductionLoops() == 0)
+    return detail::MatchContractionResult::NoReduction;
+  if (llvm::any_of(mapRange,
+                   [](AffineMap m) { return !m.isProjectedPermutation(); }))
+    return detail::MatchContractionResult::NotProjectedPermutations;
+  // TODO: more fields than add/mul.
+  // clang-format off
+  if (!isScaledContractionBody<
+        arith::MulFOp, arith::AddFOp,
+        arith::MulIOp, arith::AddIOp,
+        complex::MulOp, complex::AddOp,
+        arith::AndIOp, arith::OrIOp>(
+      *linalgOp.getBlock())) {
+    return detail::MatchContractionResult::NotAddMul;
+  }
+  // clang-format on
+  if (dimensions) {
+    FailureOr<ScaledContractionDimensions> res =
+        inferScaledContractionDims(linalgOp);
+    assert(succeeded(res) && "unexpected failure to infer contraction dims");
+    *dimensions = *res;
+  }
+  return detail::MatchContractionResult::Success;
+}
+
+bool isaScaledContractionOpInterface(linalg::LinalgOp linalgOp) {
+  if (!linalgOp)
+    return false;
+  Operation *op = linalgOp.getOperation();
+  return isScaledContractionImpl(op) == detail::MatchContractionResult::Success;
+}
+
+}; // namespace mlir::iree_compiler::IREE::LinalgExt
```

Comment:
I'm just going to trust that this is a correct generalization, since I assume it works empiracally.